### PR TITLE
Try to fix 1716 - internal proxy under load

### LINKED ISSUE #1716: Internal proxy resets connection from layer when running shell with mirrord and piping input
### Bug Description

When running a shell (was observed with both sh and zsh) that reads some commands from stdin with mirrord the internal proxy sometimes resets an incoming connection from a layer and an error is displayed. 

### Steps to Reproduce

```
echo "nvm use" | MIRRORD_PROGRESS_MODE=off mirrord exec -f .mirrord/mirrord.json zsh -- -is
```

where `.mirrord/mirrord.json` is:
```
{
    "accept_invalid_certificates": true,
    "target": {
        "path": {
            "deployment": "dotenv"
        }
    },
    "feature": {
        "network": {
            "incoming": "mirror",
            "outgoing": true
        },
        "fs": {
            "mode": "localwithoverrides",
            "read_only": ".*/.env"
        },
        "env": true
    }
}
```

Also reproduced with just `echo "nvm use" | MIRRORD_PROGRESS_MODE=off mirrord exec zsh -- -is`, but less often.

### Backtrace

_No response_

### Relevant Logs

```
2023-07-24T20:16:49.964783Z ERROR ThreadId(09) mirrord_layer::connection: Couldn't connect to internal proxy: Os { code: 54, kind: ConnectionReset, message: "Connection reset by peer" }, 127.0.0.1:51068                                                                                      
```
### Your operating system and version

macOS 13.4

### Local process

zsh

### Local process version

_No response_

### Additional Info

The error never reproduced with `trace` level logs.

Running zsh with that stdin results in **hundreds** of `execve`s, each of them results in the initialization of a new layer, each of those layers connects to the internal proxy (most of them soon disconnect as their process is replaced by the next executed process).

So one thing I'm looking into is whether the internal proxy is somehow overwhelmed by that crazy chain of process execusions.

### PR DESCRIPTION
**Pull request by aviramha**
Tries to fix #1716 

Try to resolve an issue where internal proxy is under heavy load since bash scripts does a lot of fork/exec by:
1. Increasing internal proxy's listen backlog (might not help on macOS)
2. Change internal proxy to create the upstream (agent) connection in a different task, allowing it to keep accepting.

I also tried
Not blocking the process we load into until we get the connection - that happens in the background. - That lead to e2e/integration failing, and we don't really need it.

### CODE REVIEW

**Comment by aviramha** [comment]
> Looks good. Let's also make the layer -> proxy connection lazy. I'll open a separate issue for that.

I tried making it semi-lazy (have it connect in the background) and it failed the tests, okay with having a backlog issue for that.

File: mirrord/cli/src/internal_proxy.rs [hunk 1]
```diff
@@ -205,8 +230,13 @@ pub(crate) async fn proxy(args: InternalProxyArgs) -> Result<()> {
             res = listener.accept() => {
                 match res {
                     Ok((stream, _)) => {
-                        let (agent_connection, _) = connect_and_ping(&config).await?;
-                        active_connections.spawn(connection_task(stream, agent_connection));
+                        let config = config.clone();
+                        active_connections.spawn(async move {
+                            match connect_and_ping(&config).await {
+                                Ok((agent_connection, _)) => connection_task(stream, agent_connection).await,
+                                Err(err) => error!("connectino to agent failed {err:#?}")
```
Comment (t4lz) [inline mirrord/cli/src/internal_proxy.rs:237]:
```suggestion
                                Err(err) => error!("connection to agent failed {err:#?}")
```

**Comment by t4lz** [comment]
The pause test also fails locally, when run alone. So it seems this introduces (or uncovers) a real issue with pause or with its test. Working to find the issue.

**Comment by aviramha** [comment]
The issue it introduced was me forgetting to set `nonblocking` on the self-created socket before converting it into tokio's :(

**Review by eyalb181 — APPROVED** [review]

**Pull request merged**